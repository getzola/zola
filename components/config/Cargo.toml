[package]
name = "config"
version = "0.1.0"
edition = "2024"
include = ["src/**/*"]

[dependencies]
serde = { version = "1.0", features = ["derive"] }

errors = { path = "../errors" }
utils = { path = "../utils" }
libs = { path = "../libs" }
