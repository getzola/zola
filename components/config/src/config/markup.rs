use std::{path::Path, sync::Arc};

use libs::syntect::{
    highlighting::{Theme, ThemeSet},
    html::css_for_theme_with_class_style,
    parsing::{SyntaxSet, SyntaxSetBuilder},
};
use serde::{Deserialize, Serialize};

use errors::{bail, Result};
use utils::types::InsertAnchor;

use crate::highlighting::{CLASS_STYLE, THEME_SET};

pub const DEFAULT_HIGHLIGHT_THEME: &str = "base16-ocean-dark";

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct ThemeCss {
    /// Which theme are we generating the CSS from
    pub theme: String,
    /// In which file are we going to output the CSS
    pub filename: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(default)]
pub struct Markdown {
    /// Whether to highlight all code blocks found in markdown files. Defaults to false
    pub highlight_code: bool,
    /// Emit an error for missing highlight languages. Defaults to false
    pub error_on_missing_highlight: bool,
    /// Which themes to use for code highlighting. See Readme for supported themes
    /// Defaults to "base16-ocean-dark"
    pub highlight_theme: String,
    /// Generate CSS files for Themes out of syntect
    pub highlight_themes_css: Vec<ThemeCss>,
    /// Whether to render emoji aliases (e.g.: :smile: => ðŸ˜„) in the markdown files
    pub render_emoji: bool,
    /// CSS class to add to external links
    pub external_links_class: Option<String>,
    /// Whether external links are to be opened in a new tab
    /// If this is true, a `rel="noopener"` will always automatically be added for security reasons
    pub external_links_target_blank: bool,
    /// Whether to set rel="nofollow" for all external links
    pub external_links_no_follow: bool,
    /// Whether to set rel="noreferrer" for all external links
    pub external_links_no_referrer: bool,
    /// Whether smart punctuation is enabled (changing quotes, dashes, dots etc in their typographic form)
    pub smart_punctuation: bool,
    /// Whether parsing of definition lists is enabled
    pub definition_list: bool,
    /// Whether footnotes are rendered at the bottom in the style of GitHub.
    pub bottom_footnotes: bool,
    /// A list of directories to search for additional `.sublime-syntax` and `.tmTheme` files in.
    pub extra_syntaxes_and_themes: Vec<String>,
    /// The compiled extra syntaxes into a syntax set
    #[serde(skip_serializing, skip_deserializing)] // not a typo, 2 are need
    pub extra_syntax_set: Option<SyntaxSet>,
    /// The compiled extra themes into a theme set
    #[serde(skip_serializing, skip_deserializing)] // not a typo, 2 are need
    pub extra_theme_set: Arc<Option<ThemeSet>>,
    /// Add loading="lazy" decoding="async" to img tags. When turned on, the alt text must be plain text. Defaults to false
    pub lazy_async_image: bool,
    /// Whether to insert a link for each header like the ones you can see in this site if you hover one
    /// The default template can be overridden by creating a `anchor-link.html` in the `templates` directory
    pub insert_anchor_links: InsertAnchor,
    /// Whether to enable GitHub-style alerts
    pub github_alerts: bool,
}

impl Markdown {
    pub fn validate_external_links_class(&self) -> Result<()> {
        // Validate external link class doesn't contain quotes which would break HTML and aren't valid in CSS
        if let Some(class) = &self.external_links_class {
            if class.contains('"') || class.contains('\'') {
                bail!("External link class '{}' cannot contain quotes", class)
            }
        }
        Ok(())
    }

    /// Gets the configured highlight theme from the THEME_SET or the config's extra_theme_set
    /// Returns None if the configured highlighting theme is set to use css
    pub fn get_highlight_theme(&self) -> Option<&Theme> {
        if self.highlight_theme == "css" {
            None
        } else {
            self.get_highlight_theme_by_name(&self.highlight_theme)
        }
    }

    /// Gets an arbitrary theme from the THEME_SET or the extra_theme_set
    pub fn get_highlight_theme_by_name(&self, theme_name: &str) -> Option<&Theme> {
        (*self.extra_theme_set)
            .as_ref()
            .and_then(|ts| ts.themes.get(theme_name))
            .or_else(|| THEME_SET.themes.get(theme_name))
    }

    /// Attempt to load any extra syntaxes and themes found in the extra_syntaxes_and_themes folders
    pub fn load_extra_syntaxes_and_highlight_themes(
        &self,
        base_path: &Path,
    ) -> Result<(Option<SyntaxSet>, Option<ThemeSet>)> {
        if self.extra_syntaxes_and_themes.is_empty() {
            return Ok((None, None));
        }

        let mut ss = SyntaxSetBuilder::new();
        let mut ts = ThemeSet::new();
        for dir in &self.extra_syntaxes_and_themes {
            ss.add_from_folder(base_path.join(dir), true)?;
            ts.add_from_folder(base_path.join(dir))?;
        }
        let ss = ss.build();

        Ok((
            if ss.syntaxes().is_empty() { None } else { Some(ss) },
            if ts.themes.is_empty() { None } else { Some(ts) },
        ))
    }

    pub fn export_theme_css(&self, theme_name: &str) -> Result<String> {
        if let Some(theme) = self.get_highlight_theme_by_name(theme_name) {
            Ok(css_for_theme_with_class_style(theme, CLASS_STYLE)
                .expect("the function can't even error?"))
        } else {
            bail!("Theme {} not found", theme_name)
        }
    }

    pub fn init_extra_syntaxes_and_highlight_themes(&mut self, path: &Path) -> Result<()> {
        let (loaded_extra_syntaxes, loaded_extra_highlight_themes) =
            self.load_extra_syntaxes_and_highlight_themes(path)?;

        if let Some(extra_syntax_set) = loaded_extra_syntaxes {
            self.extra_syntax_set = Some(extra_syntax_set);
        }

        if let Some(extra_theme_set) = loaded_extra_highlight_themes {
            self.extra_theme_set = Arc::new(Some(extra_theme_set));
        }

        if self.highlight_theme == "css" {
            return Ok(());
        }

        // Validate that the chosen highlight_theme exists in the loaded highlight theme sets
        if !THEME_SET.themes.contains_key(&self.highlight_theme) {
            if let Some(extra) = &*self.extra_theme_set {
                if !extra.themes.contains_key(&self.highlight_theme) {
                    bail!(
                        "Highlight theme {} not found in the extra theme set",
                        self.highlight_theme
                    )
                }
            } else {
                bail!(
                    "Highlight theme {} not available.\n\
                You can load custom themes by configuring `extra_syntaxes_and_themes` to include a list of folders containing '.tmTheme' files",
                    self.highlight_theme
                )
            }
        }

        // Validate that all exported highlight themes exist as well
        for theme in self.highlight_themes_css.iter() {
            let theme_name = &theme.theme;
            if !THEME_SET.themes.contains_key(theme_name) {
                // Check extra themes
                if let Some(extra) = &*self.extra_theme_set {
                    if !extra.themes.contains_key(theme_name) {
                        bail!(
                            "Can't export highlight theme {}, as it does not exist.\n\
                        Make sure it's spelled correctly, or your custom .tmTheme' is defined properly.",
                            theme_name
                        )
                    }
                }
            }
        }

        Ok(())
    }

    pub fn has_external_link_tweaks(&self) -> bool {
        self.external_links_target_blank
            || self.external_links_no_follow
            || self.external_links_no_referrer
            || self.external_links_class.is_some()
    }

    pub fn construct_external_link_tag(&self, url: &str, title: &str) -> String {
        let mut rel_opts = Vec::new();
        let mut target = "".to_owned();
        let title = if title.is_empty() { "".to_owned() } else { format!("title=\"{}\" ", title) };

        let class = self
            .external_links_class
            .as_ref()
            .map_or("".to_owned(), |c| format!("class=\"{}\" ", c));

        if self.external_links_target_blank {
            // Security risk otherwise
            rel_opts.push("noopener");
            target = "target=\"_blank\" ".to_owned();
        }
        if self.external_links_no_follow {
            rel_opts.push("nofollow");
        }
        if self.external_links_no_referrer {
            rel_opts.push("noreferrer");
        }
        let rel = if rel_opts.is_empty() {
            "".to_owned()
        } else {
            format!("rel=\"{}\" ", rel_opts.join(" "))
        };

        format!("<a {}{}{}{}href=\"{}\">", class, rel, target, title, url)
    }
}

impl Default for Markdown {
    fn default() -> Markdown {
        Markdown {
            highlight_code: false,
            error_on_missing_highlight: false,
            highlight_theme: DEFAULT_HIGHLIGHT_THEME.to_owned(),
            highlight_themes_css: Vec::new(),
            render_emoji: false,
            external_links_class: None,
            external_links_target_blank: false,
            external_links_no_follow: false,
            external_links_no_referrer: false,
            smart_punctuation: false,
            definition_list: false,
            bottom_footnotes: false,
            extra_syntaxes_and_themes: vec![],
            extra_syntax_set: None,
            extra_theme_set: Arc::new(None),
            lazy_async_image: false,
            insert_anchor_links: InsertAnchor::None,
            github_alerts: false,
        }
    }
}
