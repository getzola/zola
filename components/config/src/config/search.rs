use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum IndexFormat {
    ElasticlunrJson,
    #[default]
    ElasticlunrJavascript,
    FuseJson,
    FuseJavascript,
}

impl IndexFormat {
    /// file extension which ought to be used for this index format.
    fn extension(&self) -> &'static str {
        match *self {
            IndexFormat::ElasticlunrJavascript | IndexFormat::FuseJavascript => "js",
            IndexFormat::ElasticlunrJson | IndexFormat::FuseJson => "json",
        }
    }

    /// the filename which ought to be used for this format and language `lang`
    pub fn filename(&self, lang: &str) -> String {
        format!("search_index.{}.{}", lang, self.extension())
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct Search {
    /// Include the title of the page in the search index. `true` by default.
    pub include_title: bool,
    /// Includes the whole content in the search index. Ok for small sites but becomes
    /// too big on large sites. `true` by default.
    pub include_content: bool,
    /// Optionally truncate the content down to `n` code points. This might cut content in a word
    pub truncate_content_length: Option<usize>,
    /// Includes the description in the search index. When the site becomes too large, you can switch
    /// to that instead. `false` by default
    pub include_description: bool,
    /// Include the RFC3339 datetime of the page in the search index. `false` by default.
    pub include_date: bool,
    /// Include the path of the page in the search index. `false` by default.
    pub include_path: bool,
    /// Format of the search index to be produced. 'elasticlunr_javascript' by default.
    pub index_format: IndexFormat,
}

impl Default for Search {
    fn default() -> Self {
        Search {
            include_title: true,
            include_content: true,
            include_description: false,
            include_path: false,
            include_date: false,
            truncate_content_length: None,
            index_format: Default::default(),
        }
    }
}

impl Search {
    pub fn serialize(&self) -> SerializedSearch<'_> {
        SerializedSearch { index_format: &self.index_format }
    }
}

#[derive(Serialize)]
pub struct SerializedSearch<'a> {
    pub index_format: &'a IndexFormat,
}
