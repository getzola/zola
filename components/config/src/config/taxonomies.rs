use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct TaxonomyConfig {
    /// The name used in the URL, usually the plural
    pub name: String,
    /// The prefix used in the URL.
    pub prefix: Option<String>,
    /// The slug according to the config slugification strategy
    pub slug: String,
    /// If this is set, the list of individual taxonomy term page will be paginated
    /// by this much
    pub paginate_by: Option<usize>,
    pub paginate_path: Option<String>,
    /// Whether the taxonomy will be rendered, defaults to `true`
    pub render: bool,
    /// Whether to generate a feed only for each taxonomy term, defaults to `false`
    pub feed: bool,
}

impl Default for TaxonomyConfig {
    fn default() -> Self {
        Self {
            name: String::new(),
            prefix: None,
            slug: String::new(),
            paginate_by: None,
            paginate_path: None,
            render: true,
            feed: false,
        }
    }
}

impl TaxonomyConfig {
    pub fn is_paginated(&self) -> bool {
        self.paginate_by.is_some_and(|paginate_by| paginate_by > 0)
    }

    pub fn paginate_path(&self) -> &str {
        self.paginate_path.as_deref().unwrap_or("page")
    }
}
