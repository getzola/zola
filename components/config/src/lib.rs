mod config;
pub mod highlighting;
mod theme;

use std::path::Path;

pub use crate::config::{
    languages::LanguageOptions,
    link_checker::LinkChecker,
    link_checker::LinkCheckerLevel,
    markup::MathRendering,
    search::{IndexFormat, Search},
    slugify::Slugify,
    taxonomies::TaxonomyConfig,
    Config,
};
use errors::Result;

/// Get and parse the config.
/// If it doesn't succeed, exit
pub fn get_config(filename: &Path) -> Result<Config> {
    Config::from_file(filename)
}
