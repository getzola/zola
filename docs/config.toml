base_url = "https://www.getzola.org/"
title = "Zola"
description = "Everything you need to make a static site engine in one binary."

compile_sass = true
build_search_index = true

taxonomies = [
  { name = "theme-tags"}
]

[search]
index_format = "elasticlunr_json"

[markdown]
highlight_code = true
highlight_theme = "kronuz"
external_links_class = "external"
#highlight_theme = "css"
#highlight_themes_css = [
#  { theme = "base16-ocean-dark", filename = "syntax-theme-dark.css" },
#  { theme = "base16-ocean-light", filename = "syntax-theme-light.css" },
#]

[link_checker]
internal_level = "warn"

[extra]
author = "Vincent Prouillet"
