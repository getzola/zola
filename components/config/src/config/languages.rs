use std::collections::HashMap;

use errors::{Result, bail};
use libs::unic_langid::LanguageIdentifier;
use serde::{Deserialize, Serialize};

use crate::config::search;
use crate::config::taxonomies;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct LanguageOptions {
    /// Title of the site. Defaults to None
    pub title: Option<String>,
    /// Description of the site. Defaults to None
    pub description: Option<String>,
    /// Whether to generate feeds for that language, defaults to `false`
    pub generate_feeds: bool,
    /// The filenames to use for feeds. Used to find the templates, too.
    /// Defaults to ["atom.xml"], with "rss.xml" also having a template provided out of the box.
    pub feed_filenames: Vec<String>,
    pub taxonomies: Vec<taxonomies::TaxonomyConfig>,
    /// Whether to generate search index for that language, defaults to `false`
    pub build_search_index: bool,
    /// The search config, telling what to include in the search index for that language
    pub search: search::Search,
    /// A toml crate `Table` with String key representing term and value
    /// another `String` representing its translation.
    /// Use `get_translation()` method for translating key into different languages.
    pub translations: HashMap<String, String>,
}

impl LanguageOptions {
    /// Merges self with another LanguageOptions, erroring if 2 equivalent fields are not None,
    /// empty or the default value.
    pub fn merge(&mut self, other: &LanguageOptions) -> Result<()> {
        macro_rules! merge_field {
            ($orig_field:expr,$other_field:expr,$name:expr) => {
                match &$orig_field {
                    None => $orig_field = $other_field.clone(),
                    Some(cur_value) => {
                        if let Some(other_field_value) = &$other_field {
                            bail!(
                                "`{}` for default language is specified twice, as {:?} and {:?}.",
                                $name,
                                cur_value,
                                other_field_value
                            );
                        }
                    }
                };
            };
            ($cond:expr,$orig_field:expr,$other_field:expr,$name:expr) => {
                if $cond {
                    $orig_field = $other_field.clone();
                } else if !$other_field.is_empty() {
                    bail!(
                        "`{}` for default language is specified twice, as {:?} and {:?}.",
                        $name,
                        $orig_field,
                        $other_field
                    )
                }
            };
        }
        merge_field!(self.title, other.title, "title");
        merge_field!(self.description, other.description, "description");
        merge_field!(
            self.feed_filenames.is_empty()
                || self.feed_filenames == LanguageOptions::default().feed_filenames,
            self.feed_filenames,
            other.feed_filenames,
            "feed_filename"
        );
        merge_field!(self.taxonomies.is_empty(), self.taxonomies, other.taxonomies, "taxonomies");
        merge_field!(
            self.translations.is_empty(),
            self.translations,
            other.translations,
            "translations"
        );

        self.generate_feeds = self.generate_feeds || other.generate_feeds;
        self.build_search_index = self.build_search_index || other.build_search_index;

        if self.search == search::Search::default() {
            self.search = other.search.clone();
        } else if self.search != other.search {
            bail!(
                "`search` for default language is specified twice, as {:?} and {:?}.",
                self.search,
                other.search
            );
        }

        Ok(())
    }
}

impl Default for LanguageOptions {
    fn default() -> LanguageOptions {
        LanguageOptions {
            title: None,
            description: None,
            generate_feeds: false,
            feed_filenames: vec!["atom.xml".to_string()],
            taxonomies: vec![],
            build_search_index: false,
            search: search::Search::default(),
            translations: HashMap::new(),
        }
    }
}

/// We want to ensure the language codes are valid ones
pub fn validate_code(code: &str) -> Result<()> {
    if LanguageIdentifier::from_bytes(code.as_bytes()).is_err() {
        bail!(
            "Language `{}` is not a valid Unicode Language Identifier (see http://unicode.org/reports/tr35/#Unicode_language_identifier)",
            code
        )
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn merge_without_conflict() {
        let mut base_default_language_options = LanguageOptions {
            title: Some("Site's title".to_string()),
            description: None,
            generate_feeds: true,
            feed_filenames: vec!["atom.xml".to_string()],
            taxonomies: vec![],
            build_search_index: true,
            search: search::Search::default(),
            translations: HashMap::new(),
        };

        let section_default_language_options = LanguageOptions {
            title: None,
            description: Some("Site's description".to_string()),
            generate_feeds: false,
            feed_filenames: vec!["rss.xml".to_string()],
            taxonomies: vec![],
            build_search_index: true,
            search: search::Search::default(),
            translations: HashMap::new(),
        };

        base_default_language_options.merge(&section_default_language_options).unwrap();
    }

    #[test]
    fn merge_with_conflict() {
        let mut base_default_language_options = LanguageOptions {
            title: Some("Site's title".to_string()),
            description: Some("Duplicate site description".to_string()),
            generate_feeds: true,
            feed_filenames: vec![],
            taxonomies: vec![],
            build_search_index: true,
            search: search::Search::default(),
            translations: HashMap::new(),
        };

        let section_default_language_options = LanguageOptions {
            title: None,
            description: Some("Site's description".to_string()),
            generate_feeds: false,
            feed_filenames: vec!["Some feed_filename".to_string()],
            taxonomies: vec![],
            build_search_index: true,
            search: search::Search::default(),
            translations: HashMap::new(),
        };

        let res =
            base_default_language_options.merge(&section_default_language_options).unwrap_err();
        assert!(res.to_string().contains("`description` for default language is specified twice"));
    }
}
