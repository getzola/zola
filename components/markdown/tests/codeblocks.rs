use config::Config;

mod common;

enum HighlightMode {
    None,
    Inlined,
    Classed,
}

fn render_codeblock(content: &str, highlight_mode: HighlightMode) -> String {
    let mut config = Config::default_for_test();
    match highlight_mode {
        HighlightMode::None => {}
        HighlightMode::Inlined => {
            config.markdown.highlight_code = true;
        }
        HighlightMode::Classed => {
            config.markdown.highlight_code = true;
            config.markdown.highlight_theme = "css".to_owned();
        }
    }
    common::render_with_config(content, config).unwrap().body
}

#[test]
fn does_nothing_with_highlighting_disabled() {
    let body = render_codeblock(
        r#"
```
foo
bar
```
    "#,
        HighlightMode::None,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_hide_lines() {
    let body = render_codeblock(
        r#"
```hide_lines=2
foo
bar
baz
bat
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_single_line() {
    let body = render_codeblock(
        r#"
```hl_lines=2
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_line_range() {
    let body = render_codeblock(
        r#"
```hl_lines=2-3
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_all_lines() {
    let body = render_codeblock(
        r#"
```hl_lines=1-4
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_zero_start_same_as_one() {
    let body = render_codeblock(
        r#"
```hl_lines=0-3
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    let body2 = render_codeblock(
        r#"
```hl_lines=1-3
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    assert_eq!(body, body2);
}

#[test]
fn can_highlight_at_end() {
    let body = render_codeblock(
        r#"
```hl_lines=3-4
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_out_of_bounds() {
    let body = render_codeblock(
        r#"
```hl_lines=3-4567898765
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_ranges_overlap() {
    let body = render_codeblock(
        r#"
```hl_lines=2-3 1-2
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_weird_fence_tokens() {
    let body = render_codeblock(
        r#"
```hl_lines=2-3,   hl_lines      = 1 - 2
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_mix_line_ranges() {
    let body = render_codeblock(
        r#"
```hl_lines=1 3-4
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_single_line_range() {
    let body = render_codeblock(
        r#"
```hl_lines=2-2
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_reversed_range() {
    let body = render_codeblock(
        r#"
```hl_lines=3-2
foo
bar
bar
baz
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_with_classes() {
    let body = render_codeblock(
        r#"
```html,hl_lines=3-4
<link
    rel="stylesheet"
    type="text/css"
    href="main.css"
/>
```
    "#,
        HighlightMode::Classed,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_add_line_numbers() {
    let body = render_codeblock(
        r#"
```linenos
foo
bar
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_add_line_numbers_windows_eol() {
    let body = render_codeblock("```linenos\r\nfoo\r\nbar\r\n```\r\n", HighlightMode::Inlined);
    insta::assert_snapshot!(body);
}

#[test]
fn can_add_line_numbers_with_lineno_start() {
    let body = render_codeblock(
        r#"
```linenos, linenostart=40
foo
bar
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_add_line_numbers_with_highlight() {
    let body = render_codeblock(
        r#"
```linenos, hl_lines=2
foo
bar
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_add_line_numbers_with_classes() {
    let body = render_codeblock(
        r#"
```html,linenos
<link
    rel="stylesheet"
    type="text/css"
    href="main.css"
/>
```
    "#,
        HighlightMode::Classed,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_render_shortcode_in_codeblock() {
    let body = render_codeblock(
        r#"
```html,linenos
<div id="custom-attr">
{{ out_put_id(id="dQw4w9WgXcQ") }}
</div>
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_render_multiple_shortcodes_in_codeblock() {
    let body = render_codeblock(
        r#"
```linenos
text1
{{ out_put_id(id="first") }}
text2
{{ out_put_id(id="second") }}
text3
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_render_completely_mixed_codeblock() {
    let body = render_codeblock(
        r#"
```html,linenos
<a href="javascript:void(0);">{{/* before(texts="1") */}}</a>
Normally people would not write something & like <> thisï¼š
<div id="custom-attr">
An inline {{ out_put_id(id="dQw4w9WgXcQ") }} shortcode
</div>
Plain text in-between
{%/* quote(author="Vincent") */%}
A quote
{%/* end */%}
{# A Tera comment, you should see it #}
<!-- end text goes here -->
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_overflow_scroll() {
    let body = render_codeblock(
        r#"
```html,overflow_scroll
<p>Example of a very long line of source code in HTML, which will trigger a scrollbar to appear on the rendered codeblock instead of it overflowing</p>
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}

#[test]
fn can_highlight_unknown_lang() {
    let body = render_codeblock(
        r#"
```rustscript
foo
bar
```
    "#,
        HighlightMode::Inlined,
    );
    insta::assert_snapshot!(body);
}
