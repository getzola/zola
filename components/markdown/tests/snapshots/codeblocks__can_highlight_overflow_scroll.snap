---
source: components/markdown/tests/codeblocks.rs
assertion_line: 390
expression: body
---
<pre data-lang="html" style="background-color:#2b303b;color:#c0c5ce;overflow: scroll;" class="language-html "><code class="language-html" data-lang="html"><span>&lt;</span><span style="color:#bf616a;">p</span><span>&gt;Example of a very long line of source code in HTML, which will trigger a scrollbar to appear on the rendered codeblock instead of it overflowing&lt;/</span><span style="color:#bf616a;">p</span><span>&gt;
</span></code></pre>
